walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```
@@ -14,10 +14,83 @@
+  public static final int START_OF_DAY = TimeRange.getTimeInMinutes(0, 0);
+  public static final int END_OF_DAY = TimeRange.getTimeInMinutes(23, 59);
+
+  public static final Comparator<Event> ORDER_BY_START =
+      new Comparator<Event>() {
+        @Override
+        public int compare(Event a, Event b) {
+          return Long.compare(a.getWhen().start(), b.getWhen().start());
+        }
+      };
+
+  public Collection<TimeRange> query(Collection<Event> eventsCollection, MeetingRequest request) {
+    HashSet<String> requestAttendees = new HashSet<String>(request.getAttendees());
+    ArrayList<Event> events =
+        removeIrrelevantEvents(requestAttendees, new ArrayList<Event>(eventsCollection));
+    List<TimeRange> res = new ArrayList<TimeRange>();
+    if (events.isEmpty()) {
+      addIfValid(TimeRange.fromStartEnd(0, END_OF_DAY, true), res, request.getDuration());
+      return res;
+    }
+
+    Collections.sort(events, ORDER_BY_START);
+
+    // Add first gap.
+    addIfValid(
+        TimeRange.fromStartEnd(0, events.get(0).getWhen().start(), false),
+        res,
+        request.getDuration());
+    int end = events.get(0).getWhen().end();
+    for (Event event : events) {
+      // event can be merged with current time range
+      if (event.getWhen().start() <= end) {
+        end = Math.max(end, event.getWhen().end());
+      } else {
+        // Add the time range we were tracking, start a new one from event.
+        addIfValid(
+            TimeRange.fromStartEnd(end, event.getWhen().start(), false),
+            res,
+            request.getDuration());
+        end = event.getWhen().end();
+      }
+    }
+
+    // Add the last one we were tracking.
+    addIfValid(TimeRange.fromStartEnd(end, END_OF_DAY, true), res, request.getDuration());
+    return res;
+  }
+
+  void addIfValid(TimeRange range, List<TimeRange> ranges, long duration) {
+    if (range.duration() >= duration) {
```

Is silently dropping "invalid" ranges really the behavior you want?

It might be, but I suspect this will result in bad user experience for you API, because there's no obvious indication when input is being dropped. Consider throwing an exception instead.

---

lt adds range to ranges if it at least duration time long (if a meeting can be scheduled in range). l'll add a comment and change the name accordingly.
But this is the behavior l want.